client/app/cloud/project/compute/infrastructure/virtualMachine/add/cloud-project-compute-infrastructure-virtualMachine-add.controller.js

```diff
@@ -0,0 +1,376 @@
+                    filteredFlavors = _.filter(filteredFlavors, flavor => _.indexOf(restrictedFlavors, flavor.shortType) > -1);
+                }
+
+                this.groupedFlavors = this.VirtualMachineAddService.groupFlavorsByCategory(filteredFlavors, this.enums.flavorsTypes);
+            })
+            .catch(this.ServiceHelper.errorHandler("cpcivm_add_step3_flavors_ERROR"))
+            .finally(() => {
+                this.loaders.step3 = false;
+            });
+    }
+
+    isStep3Valid () {
+        return this.model.flavor != null && !_.isEmpty(this.model.name) && this.model.number > 0 && (!this.state.hasVRack || !_.isEmpty(this.model.networkId));
+    }
+
+    resetStep3 () {
+        _.set(this.model, "flavor", null);
+        _.set(this.model, "network", null);
+        _.set(this.model, "number", 1);
+        if (!this.isNameUpdated) {
+            _.set(this.model, "name", "");
+        }
+        this.resetStep4();
+    }
+
+    enablePostScript () {
+        this.isPostScriptEnabled = true;
+    }
+
+    getPrivateNetworks () {
+        _.set(this.loaders, "privateNetworks", true);
+        return this.OvhApiCloudProjectNetworkPrivate.Lexi().query({ serviceName: this.serviceName }).$promise.then(networks => {
+            this.privateNetworks = networks;
+            return this.VirtualMachineAddService.getPrivateNetworksSubNets(this.serviceName, this.privateNetworks);
+        }).then(subNets => {
+            this.displayedPrivateNetworks = this.VirtualMachineAddService.getFilteredPrivateNetworksByRegion(this.privateNetworks, this.model.region.microRegion.code, subNets);
+        }).catch(() => {
+            this.displayedPrivateNetworks = [];
+        }).finally(() => {
+            this.loaders.privateNetworks = false;
+        });
+    }
+
+    setInstanceName () {
+        if (_.isEmpty(this.model.name) || !this.isNameUpdated) {
+            this.model.name = `${_.get(this.model, "flavor.name", "")}-${_.get(this.model, "region.microRegion.code", "")}`.toLowerCase();
+        }
+    }
+
+    /*--------------------------
+     |  Step 4: Billing period  |
+     --------------------------*/
+
+    isStep4Valid () {
+        return _.isString(this.model.billingPeriod) && !_.isEmpty(this.model.billingPeriod);
+    }
+
+    resetStep4 () {
+        _.set(this.model, "billingPeriod", null);
+    }
+
+    /*-------------------
+     |  Submit the form  |
+     -------------------*/
+
+    addVirtualMachine () {
+        this.loaders.adding = true;
+
+        if (!_.isEmpty(this.model.networkId) && this.model.networkId !== "none") {
+            this.model.networks = [{ networkId: this.model.networkId }, { networkId: _.first(this.publicNetworks).id }];
+        }
+
+        return this.VirtualMachineAddService.createVirtualMachine(this.serviceName, this.model)
+            .then(() => {
+                this.previousState.go();
+            }).catch(this.ServiceHelper.errorHandler("cpcivm_add_launch_ERROR")).finally(() => {
```

Comment:
its nothing important but could you put finally on a new line also?

Reply:
done